### PR TITLE
Hive: Avoid recursive listing in HiveCatalog#renameTable

### PR DESCRIPTION
**Pull request by singhpk234**
Attempts to avoid recursive listing in HiveCatalog#renameTable

ref : https://github.com/apache/iceberg/pull/734#issuecomment-574062407

--- 

Closes https://github.com/apache/iceberg/issues/4406

### CODE REVIEW

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 1]
```diff
@@ -64,6 +68,14 @@
 public class HiveCatalog extends BaseMetastoreCatalog implements SupportsNamespaces, Configurable {
   public static final String LIST_ALL_TABLES = "list-all-tables";
   public static final String LIST_ALL_TABLES_DEFAULT = "false";
+  private static final DynMethods.UnboundMethod ALTER_TABLE = DynMethods.builder("alter_table")
```
Comment (RussellSpitzer) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:71]: I did kinda wonder if we should extract this to a package utility method ? But this seems fine with me if you want to go this route. It's only used in 2 places so I think this is fine

**Review by RussellSpitzer — APPROVED** [review]

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 1 above)
Comment (singhpk234) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:71]: +1, Agree with you on this, I think we can move this to the existing [MetaStoreUtil](https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java) in the package WDYT ?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java [hunk 2]
```diff
@@ -37,6 +53,17 @@ public static boolean hive3PresentOnClasspath() {
     return HIVE3_PRESENT_ON_CLASSPATH;
   }
 
+  /**
+   * Invokes alterTable via metaStoreClient without triggering recursive listing.
+   * ref: https://github.com/apache/iceberg/pull/734#issuecomment-574062407
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java:58]: Rather than referencing the issue, it would be more clear to just state the resolution: if possible, call the alter_table method with an environment context that turns off stats updates.

**Review by rdblue — CHANGES_REQUESTED** [review]
I think this is good, but should have a minor update to docs so that the change is clear without going to an issue.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java [hunk 3]
```diff
@@ -54,8 +54,8 @@ public static boolean hive3PresentOnClasspath() {
   }
 
   /**
-   * Invokes alterTable via metaStoreClient without triggering recursive listing.
-   * ref: https://github.com/apache/iceberg/pull/734#issuecomment-574062407
+   * If possible, call the alter_table method via metaStoreClient with an environment context that turns off stats
```
[the remark this answers is not in the document]
Comment (singhpk234) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java:57]: > Nits: Why "if possible"? The final comma is also not needed so

It's due to this signature which is added as final fallback 
```java
      .impl(IMetaStoreClient.class, "alter_table",
          String.class, String.class, Table.class)
```

> "The final comma is also not needed so" 

makes sense.

should I rephrase it to : 

" If possible, calls alter_table method using the metastore client with an environmental context that turns off stats update to avoid recursive file listing"

[the remark this answers is not in the document]
Comment (RussellSpitzer) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java:57]: The "if possible" here applies to the whole sentence which makes it sound like this may call alter_table or it may not. So maybe instead
"Calls alter_table method using the metastore client. If possible an environmental context will be used that ...."

**Review by RussellSpitzer — APPROVED** [review]

**Review by rdblue — APPROVED** [review]

**Comment by RussellSpitzer** [comment]
Thanks @singhpk234 for the PR @rdblue for helping out with review! Merged
